sdk/src/context.rs

```diff
@@ -74,44 +78,32 @@ where
 {
     pub fn new(
         anchor_context: Context<'a, 'b, 'c, 'info, T>,
-        inputs: Vec<Vec<u8>>,
-        merkle_context: PackedMerkleContext,
-        merkle_tree_root_index: u16,
-        address_merkle_context: PackedAddressMerkleContext,
-        address_merkle_tree_root_index: u16,
+        inputs: LightInstructionInputs,
     ) -> Result<Self> {
-        let light_accounts = U::try_light_accounts(
-            inputs,
-            merkle_context,
-            merkle_tree_root_index,
-            address_merkle_context,
-            address_merkle_tree_root_index,
-            anchor_context.remaining_accounts,
-        )?;
+        let light_accounts = U::try_light_accounts(&inputs)?;
         Ok(Self {
             anchor_context,
             light_accounts,
+            inputs,
         })
     }
 
-    pub fn verify(&mut self, proof: CompressedProof) -> Result<()> {
+    pub fn verify(&mut self) -> Result<()> {
         let bump = Pubkey::find_program_address(
             &[CPI_AUTHORITY_PDA_SEED],
             &self.anchor_context.accounts.get_invoking_program().key(),
         )
         .1;
         let signer_seeds = [CPI_AUTHORITY_PDA_SEED, &[bump]];
 
-        let new_address_params = self.light_accounts.new_address_params();
-        let input_compressed_accounts_with_merkle_context = self
-            .light_accounts
-            .input_accounts(self.anchor_context.remaining_accounts)?;
-        let output_compressed_accounts = self
-            .light_accounts
-            .output_accounts(self.anchor_context.remaining_accounts)?;
+        // let new_address_params = self.light_accounts.new_address_params();
+        let new_address_params = self.inputs.new_addresses.clone().unwrap_or(Vec::new());
+        let input_compressed_accounts_with_merkle_context =
+            self.inputs.accounts.clone().unwrap_or(Vec::new());
```

Comment:
these clones look dangerous to me

Reply:
separate serialization also opens the door to optimize PackedCompressedAccounts and their underlying data separately from the other inputs.

Reply:
in general for deserialization the following pattern would be useful so that it is easy to use in a different context and easy to extend:
`pub fn deserialize<T>(&mut start_index, data.as_slice())`

Reply:
Great idea, I will try that out. And yes, I was meaning to take care of all `clone()` calls before marking as ready for review, it's just a dirty prototype.